lib/seahorn/BvOpsem2WideMemMgr.cc

```diff
@@ -0,0 +1,543 @@
+    uint64_t gvSz = m_sem.getTD().getTypeAllocSize(gv.getValueType());
+    auto range = m_allocator->galloc(gv, gvSz, std::max(align, m_alignment));
+    return PtrTy(m_ctx.alu().si(range.first, ptrSzInBits()),
+                 bytesToSlotExpr(range.second - range.first));
+  }
+
+  PtrTy falloc(const Function &fn) {
+    auto range = m_allocator->falloc(fn, m_alignment);
+    return PtrTy(m_ctx.alu().si(range.first, ptrSzInBits()),
+                 bytesToSlotExpr(range.second - range.first));
+  }
+
+  // TODO: What is the right size to return here?
+  PtrTy getPtrToFunction(const Function &F) {
+    m_size.getPtrToFunction(F);
+    return PtrTy(m_main.getPtrToFunction(F), m_uninit_size);
+  }
+
+  PtrTy getPtrToGlobalVariable(const GlobalVariable &gv) {
+    uint64_t gvSz = m_sem.getTD().getTypeAllocSize(gv.getValueType());
+    return PtrTy(m_ctx.alu().si(
+                     m_allocator->getGlobalVariableAddr(gv, gvSz, m_alignment),
+                     ptrSzInBits()),
+                 bytesToSlotExpr(gvSz));
+  }
+
+  void initGlobalVariable(const GlobalVariable &gv) const {
+    if (!gv.hasInitializer()) {
+      LOG("opsem", WARN << "GV without an initializer: " << gv << "\n";);
+      return;
+    }
+
+    ConstantExprEvaluator ce(m_sem.getDataLayout());
+    ce.setContext(m_ctx);
+    char *bytes = m_allocator->getGlobalVariableMem(gv);
+    LOG("opsem", if (!bytes) ERR << "Unexpected nullptr storage for global: "
+                                 << gv << "\n";);
+    assert(bytes);
+    ce.initMemory(gv.getInitializer(), bytes);
+  }
+
+  PtrTy mkAlignedPtr(Expr name, uint32_t align) const {
+    m_size.mkAlignedPtr(name, align);
+    return PtrTy(m_main.mkAlignedPtr(name, align), m_uninit_size);
+  }
+
+  PtrSortTy mkPtrRegisterSort(const Instruction &inst) const {
+    return PtrSortTy(m_main.mkPtrRegisterSort(inst),
+                     m_ctx.alu().intTy(g_slotBitWidth));
+  }
+
+  PtrSortTy mkPtrRegisterSort(const Function &fn) const { return ptrSort(); }
+
+  /// \brief Returns sort of a pointer register for a global pointer
+  PtrSortTy mkPtrRegisterSort(const GlobalVariable &gv) const {
+    return ptrSort();
+  }
+
+  MemSortTy mkMemRegisterSort(const Instruction &inst) const {
+    RawMemSortTy mainSort = m_main.mkMemRegisterSort(inst);
+    RawMemSortTy sizeSort = m_size.mkMemRegisterSort(inst);
+    return MemSortTy(mainSort, sizeSort);
+  }
+
+  PtrTy freshPtr() {
+    Expr name = op::variant::variant(m_id++, m_freshPtrName);
+    return mkAlignedPtr(name, m_alignment);
+  }
+
+  PtrTy nullPtr() const { return m_nullPtr; }
+
+  Expr coerce(Expr sort, Expr val) { return m_memRepr->coerce(sort, val); }
+
+  PtrTy ptrAdd(PtrTy ptr, int32_t _offset) const {
+    if (_offset == 0)
+      return ptr;
+    expr::mpz_class offset((signed long)_offset);
+    auto address = m_ctx.alu().doAdd(
+        ptr.getRaw(), m_ctx.alu().si(offset, ptrSzInBits()), ptrSzInBits());
+    auto new_size = m_ctx.alu().doSub(
```

Comment:
optimization opportunity: `size` might be a number at this point, in which case sub can be done concretely and not symbolically

Reply:
good point. done